test/parser.go

```diff
@@ -0,0 +1,103 @@
+package main
+
+import (
+	"bufio"
+	"errors"
+	"io"
+	"strings"
+)
+
+// An Action is one instance of a START/END tag pair. ActionType is
+// the word following START. Lines are the lines between the tags.
+type Action struct {
+	ActionType string
+	Lines      []string
+	closed     bool
+}
+
+type stateFn func(string, *[]Action) stateFn
+
+func neutralState(line string, actions *[]Action) stateFn {
+	words := strings.Fields(line)
+	for index, word := range words {
+		if word == "END" {
+			// Shouldn't find an END in neutral state.
+			return nil
+		}
+		if word == "START" {
+			action := words[index+1]
+			newAction(action, actions)
+			return insideActionState
+		}
+	}
+	return neutralState
+}
+
+func insideActionState(line string, actions *[]Action) stateFn {
+	words := strings.Fields(line)
+	if strings.HasPrefix(line, "```") || words[0] == "```" {
+		// Ignore line.
+		return insideActionState
+	}
+	for _, word := range words {
+		if word == "END" {
+			closeAction(*actions)
+			return neutralState
+		}
+		if word == "START" {
+			// Shouldn't find a START inside an action.
+			return nil
+		}
+	}
+	actionLine(line, *actions)
+	return insideActionState
+}
+
+func newAction(aType string, actions *[]Action) {
+	aa := Action{
+		ActionType: aType,
+		Lines:      make([]string, 0),
+		closed:     false,
+	}
+	*actions = append(*actions, aa)
+}
+
+func actionLine(line string, actions []Action) {
+	actions[len(actions)-1].Lines =
+		append(actions[len(actions)-1].Lines, line)
+}
+
+func closeAction(actions []Action) {
+	actions[len(actions)-1].closed = true
+}
+
+// GetActions parses the input for START/END tags and compiles a list
+// of actions, comprising of the lines between the lines with
+// START/END. Lines that begin with ``` are ignored. The word
+// immediately after START is saved as the action type.
+func GetActions(reader io.Reader) ([]Action, error) {
+	scanner := bufio.NewScanner(reader)
+	actions := make([]Action, 0)
+
+	for state := neutralState; state != nil; {
+		if scanner.Scan() == false {
+			break
+		}
+		line := scanner.Text()
+		state = state(line, &actions)
+	}
```

Comment:
I'd invert this.

Check `nil` for the state, since `nil` appears to be used as an indication of error. Errors aren't reported, though?

```
	state := neutralState
	for scanner.Scan() {
		line := scanner.Text()
		state = state(line, &actions)
		if state == nil {
			break
		}
	}
```